extensions/notebook/src/book/bookTocManager.ts

```diff
@@ -51,49 +51,71 @@ export class BookTocManager implements IBookTocManager {
 		this.targetBookContentPath = targetBook?.bookItems[0].rootContentPath;
 	}
 
-	async getAllFiles(toc: JupyterBookSection[], directory: string, filesInDir: string[], rootDirectory: string): Promise<JupyterBookSection[]> {
-		await Promise.all(filesInDir.map(async file => {
-			let isDirectory = (await fs.promises.stat(path.join(directory, file))).isDirectory();
-			if (isDirectory) {
-				let files = await fs.promises.readdir(path.join(directory, file));
-				let initFile: string = '';
-				//Add files named as readme or index within the directory as the first file of the section.
-				files.some((f, index) => {
-					if (initMarkdown.includes(f)) {
-						initFile = path.parse(f).name;
-						files.splice(index, 1);
-					}
-				});
-				let jupyterSection: JupyterBookSection = {
-					title: file,
-					file: path.join(file, initFile),
-					expand_sections: true,
-					numbered: false,
-					sections: []
-				};
-				toc.push(jupyterSection);
-				await this.getAllFiles(toc, path.join(directory, file), files, rootDirectory);
-			} else if (allowedFileExtensions.includes(path.extname(file))) {
-				// if the file is in the book root we don't include the directory.
-				const filePath = directory === rootDirectory ? path.parse(file).name : path.join(path.basename(directory), path.parse(file).name);
-				const addFile: JupyterBookSection = {
-					title: path.parse(file).name,
-					file: filePath
+	/**
+	 * Replicates Jupyter Book logic to add the init markdown files
+	 * at the beginning of the table of contents.
+	 * @param files The files in the directory
+	*/
+	getInitFile(files: string[]): path.ParsedPath | undefined {
+		let initFile = undefined;
+		// Add files named as readme or index within the directory as the first file of the section.
+		files.some((f) => {
+			if (initMarkdown.includes(f.toLowerCase())) {
+				initFile = path.parse(f);
+			}
+		});
+		// If it doesnt find an initMarkdown then use the first file we find.
+		if (initFile === undefined) {
+			files.some((f) => {
+				const parsedPath = path.parse(f);
+				if (allowedFileExtensions.includes(parsedPath.ext)) {
+					initFile = parsedPath;
+				}
+			});
+		}
+		return initFile;
+	}
+
+	/**
+	 * Creates the table of contents of a book by reading the folder structure
+	 * that the user provides.
+	 * @param contents The contents of directory.
+	 * @param directory The current directory.
+	 * @param rootDirectory The root path of the book.
+	 * Returns an array of Jupyter Sections.
+	*/
+	async createTocFromDir(contents: string[], directory: string, rootDirectory: string): Promise<JupyterBookSection[]> {
+		let toc: JupyterBookSection[] = [];
+		for (const content of contents) {
```

Comment:
Are you ok with having this entire thing throw if there's any files that it can't access for some reason? Could consider just skipping those files and displaying a warning to the user...